swaps/swaps_service.go

```diff
@@ -699,31 +694,78 @@ func (svc *swapsService) RefundSwap(swapId, address string) error {
 		return err
 	}
 
-	fastestFee := float64(feeRates.FastestFee)
-	refundTransaction, _, err := boltz.ConstructTransaction(
-		network,
-		boltz.CurrencyBtc,
-		[]boltz.OutputDetails{
-			{
-				SwapId:             swapId,
-				SwapType:           boltz.NormalSwap,
-				Address:            address,
-				LockupTransaction:  lockupTransaction,
-				TimeoutBlockHeight: swapTransactionResp.TimeoutBlockHeight,
-				Vout:               vout,
-				PrivateKey:         ourKeys,
-				SwapTree:           tree,
-				Cooperative:        true,
+	waitToRetrySeconds := 0
+	var refundTransaction boltz.Transaction
+
+	for i := 0; ; i++ {
+		contextCancelled := false
+		select {
+		case <-svc.ctx.Done():
+			logger.Logger.WithField("swapId", swapId).Info("Swap refund context cancelled")
+			contextCancelled = true
+		case <-time.After(time.Duration(waitToRetrySeconds) * time.Second): // timeout
+		}
+		if contextCancelled {
+			return nil
+		}
+
+		nodeInfo, err := svc.lnClient.GetInfo(svc.ctx)
+		if err != nil {
+			logger.Logger.WithError(err).WithFields(logrus.Fields{
+				"swapId": swapId,
+			}).WithError(err).Error("Failed to request node info")
+			return err
+		}
+
+		feeRates, err := svc.getFeeRates()
+		if err != nil {
+			logger.Logger.WithField("swapId", swapId).WithError(err).Error("Failed to fetch fee rate to create claim transaction")
+			return err
+		}
+
+		cooperative := swapTransactionResp.TimeoutBlockHeight > nodeInfo.BlockHeight
+
+		fastestFee := float64(feeRates.FastestFee)
+		refundTransaction, _, err = boltz.ConstructTransaction(
+			network,
+			boltz.CurrencyBtc,
+			[]boltz.OutputDetails{
+				{
+					SwapId:             swapId,
+					SwapType:           boltz.NormalSwap,
+					Address:            address,
+					LockupTransaction:  lockupTransaction,
+					TimeoutBlockHeight: swapTransactionResp.TimeoutBlockHeight,
+					Vout:               vout,
+					PrivateKey:         ourKeys,
+					SwapTree:           tree,
+					Cooperative:        cooperative,
+				},
 			},
-		},
-		boltz.Fee{
-			SatsPerVbyte: &fastestFee,
-		},
-		svc.boltzApi,
-	)
-	if err != nil {
-		logger.Logger.WithField("swapId", swapId).WithError(err).Error("Could not create claim transaction")
-		return err
+			boltz.Fee{
+				SatsPerVbyte: &fastestFee,
+			},
+			svc.boltzApi,
+		)
+		if err != nil {
+			if enableRetries && cooperative {
+				waitToRetrySeconds = max(waitToRetrySeconds, 0)
```

Comment:
this logic for the `waitToRetrySeconds` is kind of complicated. I think we can directly calculate it based on the for loop iteration variable `i` rather than adding. And how can it even become negative?